### ISSUE TITLE
Failed to mount an External Volume Storage

### ISSUE THREAD

**Issue by mms11** [kadalu/kadalu#520]
Using Versions:
glusterfs 8.1
kubernetes v1.19.8
Kadalu 0.8.0

```
kubectl get pods -n kadalu
NAME                          READY   STATUS    RESTARTS   AGE
kadalu-csi-nodeplugin-m2r4c   3/3     Running   0          75m
kadalu-csi-nodeplugin-r6fvl   3/3     Running   0          75m
kadalu-csi-provisioner-0      5/5     Running   0          13h
operator-7fb9f49d75-tzfpm     1/1     Running   0          13h
pod-ext                       0/1     Pending   0          14h
```

My gluster storage nodes are outside k8s cluster.

```
blkid /dev/mapper/vg_xfs-xfs_db
/dev/mapper/vg_xfs-xfs_db: UUID="4b2f9589-e34d-449a-8d0f-5ea8e3c02940" TYPE="xfs"

gluster volume info kadalu-volume-test_2

Volume Name: kadalu-volume-test_2
Type: Distribute
Volume ID: d03176d8-ae90-401a-99bc-3a3dc6873f17
Status: Started
Snapshot Count: 0
Number of Bricks: 1
Transport-type: tcp
Bricks:
Brick1: 10.79.120.121:/mount/point/kadalu-volume-test
Options Reconfigured:
diagnostics.count-fop-hits: on
diagnostics.latency-measurement: on
transport.address-family: inet
nfs.disable: on

df -h | grep /mount/point
/dev/mapper/vg_xfs-xfs_db      6.0G   33M  6.0G   1% /mount/point
```

I've installed the operator that is work and running as supposed and I'm follow all the process described on "Use external gluster volume as PV in kadalu" section.

Everytime that I try to use external gluster in kadalu native way or use external storage directly I got a issues

```
$ kubectl-kadalu storage-add store-name --external 10.79.120.121:/kadalu-volume-test_2

kubectl-kadalu storage-add store-name --external 10.79.120.121:/kadalu-volume-test_2
The following nodes are available:
  vm016634, vm016635, vm016636, vm016637, vm016638, vm016639, vm016640, vm016641

Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/bin/kubectl-kadalu/__main__.py", line 74, in <module>
  File "/usr/local/bin/kubectl-kadalu/__main__.py", line 54, in main
  File "/usr/local/bin/kubectl-kadalu/storage_add.py", line 235, in run
  File "/usr/local/bin/kubectl-kadalu/storage_yaml.py", line 58, in to_storage_yaml
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/string.py", line 121, in substitute
    return self.pattern.sub(convert, self.template)
  File "/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/string.py", line 114, in convert
    return str(mapping[named])
KeyError: 'gluster_options'
```

Error on the provisioner using external storage directly
`E0422 08:43:32.002259 1 controller.go:966] error syncing claim "8b434aee-78e7-472c-a306-bb224468c683": failed to provision volume with StorageClass "kadalu.external.pv-ext": rpc error: code = InvalidArgument desc = External Storage provided not valid`

Even if I try to mount directly in POD with "mount -t glusterfs 10.79.120.121:/kadalu-volume-test_2 -o log-level=DEBUG /mount/point" it says that the mount doesnt exists.

Is it a requirement that the gluster bricks must be provided from XFS volumes?

**Comment by amarts** [comment]
XFS is not mandatory for external storage.

For kadalu native pods also XFS is not mandatory starting 0.8.0 version.

For this issue, I think there seems to a bug in `kubectl kadalu` CLI. Can you try the yaml way?

Also please check 10.79.120.121 is reachable from the host nodes?... 

One more thing to check is the logs of the operator when you use the `kadalu.external.pv-ext`!

```yaml
# file: external-config.yaml
---
apiVersion: kadalu-operator.storage/v1alpha
kind: KadaluStorage
metadata:
  name: ext-config
spec:
  type: External
  details:
    # gluster_hosts: [ gluster1.kadalu.io, gluster2.kadalu.io ]
    gluster_host: 10.79.120.121
    gluster_volname: kadalu-volume-test_2
    gluster_options: log-level=DEBUG
```

do a `kubectl apply -f <above-file>`

then check the operator logs. (`kubectl logs -nkadalu opeator*`)

now try to ask for pvc from `kadalu.external.ext-config` StorageClass.

**Comment by leelavg** [comment]
> KeyError: 'gluster_options'

Fix for above is already [merged]( https://github.com/kadalu/kadalu/commit/76869e0f5f6b0b00ab66d8825c09cbf3828b00e0#diff-c1431eee976ad389c7c8070d2069b36e7dfc85f4d9b18713f56af09fd4d86ebfR189) into `devel` and would be part of next release. 

Please follow what was suggested in above comment for now.

**Comment by mms11** [comment]
Doing that I got the follow issue
`kubectl apply -f ./external-config.yaml
error: unable to recognize "./external-config.yaml": no matches for kind "KadaluStorage" in version "kadalu-operator.storage/v1alpha"`

Meanwhile I've check the k8s api-resource and notice that correct API name is kadalu-operator.storage/v1alpha**1**
`kubectl api-resources | grep kada
kadalustorages                                 kadalu-operator.storage/v1alpha1   true         KadaluStorage`

Fixing  the kadalu native way works using kubectl apply so I'm good for now.

Do you know if is possible (using this kind of config) to use kadalu to create the volume on gluster automatically?

**Comment by leelavg** [comment]
@amarts I think recent [commit](https://github.com/kadalu/kadalu/pull/516/files) should have been documented with `v1alpha1` to correspond to src?

@mms11, 
- Thanks for bringing the version difference to our notice and thanks for the positive confirmation
- This kind of config assumes an external gluster volume already exists then ~Operator mounts that~ this volume get mounted into CSI pods to serve storage
- AFAIK, there's no plan to create external gluster volumes on demand at the moment.

**Comment by juliosibaja** [comment]
Hello!!
I have also some problems using External Volumes. I'm using a rapsberry pi 4 with:
- Raspios arm64
- Glusterfs 5.5
- One Volume (myvol1). I have tested mounting the volume in another machine with `sudo mount -t glusterfs 10.13.13.2:/myvol1 /mnt/myvol1 ` and everything worked. The volume config is the following:
``` 
Volume Name: myvol1
Type: Replicate
Volume ID: f226036d-2122-4a50-8c6c-c5ee2d70f71c
Status: Started
Snapshot Count: 0
Number of Bricks: 1 x 2 = 2
Transport-type: tcp
Bricks:
Brick1: 10.13.13.2:/mnt/myvol1/brick1/brick
Brick2: 10.13.13.3:/mnt/myvol1/brick1/brick
Options Reconfigured:
transport.address-family: inet
nfs.disable: on
performance.client-io-threads: off
``` 

I'm running K3S with kubectl v1.20.6+k3s1 and kubectl-kadalu 0.7.7
And using this config
```
apiVersion: kadalu-operator.storage/v1alpha
kind: KadaluStorage
metadata:
  name: external-configuration
spec:
  type: External
  details:
    gluster_host: 10.13.13.2
    gluster_volname: myvol1
    gluster_options: log-level=DEBUG
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: pv-ext-kadalu
spec:
  # Add 'kadalu.external.' to name from KadaluStorage kind
  storageClassName: kadalu.external.external-configuration
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      # This needs to be set using 'kadalu-quotad'
      storage: 500Mi
```
The PVC is stucked in pending and I'm getting this inside the  csi-provisioner: 
```
W0427 11:32:21.278340       1 controller.go:943] Retrying syncing claim "f33cc046-a0f7-4d14-a23e-7e647f541f38", failure 7
E0427 11:32:21.278793       1 controller.go:966] error syncing claim "f33cc046-a0f7-4d14-a23e-7e647f541f38": failed to provision volume with StorageClass "kadalu.external.external-configuration": rpc error: code = InvalidArgument desc = External Storage provided not valid
I0427 11:32:21.279374       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"kadalu", Name:"pv-ext-kadalu", UID:"f33cc046-a0f7-4d14-a23e-7e647f541f38", APIVersion:"v1", ResourceVersion:"30942", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "kadalu.external.external-configuration": rpc error: code = InvalidArgument desc = External Storage provided not valid
I0427 11:34:29.279073       1 controller.go:1317] provision "kadalu/pv-ext-kadalu" class "kadalu.external.external-configuration": started
I0427 11:34:29.279329       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"kadalu", Name:"pv-ext-kadalu", UID:"f33cc046-a0f7-4d14-a23e-7e647f541f38", APIVersion:"v1", ResourceVersion:"30942", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "kadalu/pv-ext-kadalu"
W0427 11:34:29.709691       1 controller.go:943] Retrying syncing claim "f33cc046-a0f7-4d14-a23e-7e647f541f38", failure 8
E0427 11:34:29.709791       1 controller.go:966] error syncing claim "f33cc046-a0f7-4d14-a23e-7e647f541f38": failed to provision volume with StorageClass "kadalu.external.external-configuration": rpc error: code = InvalidArgument desc = External Storage provided not valid
I0427 11:34:29.709862       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"kadalu", Name:"pv-ext-kadalu", UID:"f33cc046-a0f7-4d14-a23e-7e647f541f38", APIVersion:"v1", ResourceVersion:"30942", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "kadalu.external.external-configuration": rpc error: code = InvalidArgument desc = External Storage provided not valid
``` 
And this is inside the kadalu-provisioner
```
[2021-04-27 11:34:29,320] INFO [volumeutils - 933:mount_glusterfs_with_host] - mount command failed	 cmd=['/opt/sbin/glusterfs', '--process-name', 'fuse', '-l', '/var/log/gluster/gluster.log', '--volfile-id', 'myvol1', '--volfile-server', '10.13.13.2', '/mnt/ext-config'] error=[1]  
[2021-04-27 11:34:29,707] ERROR [controllerserver - 202:CreateVolume] - External Storage provided not valid
```
Thanks for your help

**Comment by leelavg** [comment]
@juliosibaja 

- just checked, even in 0.7.7 release `apiVersion` is `kadalu-operator.storage/v1alpha1` not `kadalu-operator.storage/v1alpha`, can you please use that while applying KadaluStorage manifest file
- i created a setup today using external gluster on devel branch which worked, if above didn't work i can only ask to recreate the cluster if you don't have any workloads running

(or)

Are you just mentioning that you already corrected the mistake and was able to use the storage?

**Comment by juliosibaja** [comment]
@leelavg Thanks for the answer. I'm sorry but i'm already using `kadalu-operator.storage/v1alpha1`, it was just a typo and yes mi PVC is always in Pending State. 
I can try with the devel branch on a new cluster installation. On the other hand as i'm using the arm version i'm not sure if i can use the [0.8.2 alpha 0 relase ](https://github.com/kadalu/kadalu/releases/tag/0.8.2alpha0) or anything else.

**Comment by juliosibaja** [comment]
I tried with a new installation of K3S using [0.8.2alpha0 relase](https://github.com/kadalu/kadalu/releases/tag/0.8.2alpha0) and also this operator: https://raw.githubusercontent.com/kadalu/kadalu/arm/manifests/kadalu-operator.yaml
I got this from `kubectl describe pvc pv-ext-kadalu`
```
Name:          pv-ext-kadalu
Namespace:     default
StorageClass:  kadalu.external.external-configuration
Status:        Pending
Volume:        
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: kadalu
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
Used By:       <none>
Events:
  Type     Reason                Age                   From                                                                  Message
  ----     ------                ----                  ----                                                                  -------
  Warning  ProvisioningFailed    20m                   kadalu_kadalu-csi-provisioner-0_89e50ae2-3a8a-4814-a845-a0e13335cedf  failed to provision volume with StorageClass "kadalu.external.external-configuration": rpc error: code = Unknown desc = Exception calling application: disk I/O error
  Normal   Provisioning          2m45s (x12 over 21m)  kadalu_kadalu-csi-provisioner-0_89e50ae2-3a8a-4814-a845-a0e13335cedf  External provisioner is provisioning volume for claim "default/pv-ext-kadalu"
  Warning  ProvisioningFailed    2m44s (x11 over 21m)  kadalu_kadalu-csi-provisioner-0_89e50ae2-3a8a-4814-a845-a0e13335cedf  failed to provision volume with StorageClass "kadalu.external.external-configuration": rpc error: code = InvalidArgument desc = External Storage provided not valid
  Normal   ExternalProvisioning  76s (x82 over 21m)    persistentvolume-controller                                           waiting for a volume to be created, either by external provisioner "kadalu" or manually created by system administrator

```
When i print the logs from the kadalu-csi-provisioner-0 i get this:
- ## csi-provisioner	
```
I0427 22:35:48.120870       1 controller.go:1317] provision "default/pv-ext-kadalu" class "kadalu.external.external-configuration": started
I0427 22:35:48.121195       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv-ext-kadalu", UID:"f4e0428a-1e14-49ac-9560-afe858a3c756", APIVersion:"v1", ResourceVersion:"3211", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "default/pv-ext-kadalu"
W0427 22:35:48.128907       1 controller.go:943] Retrying syncing claim "f4e0428a-1e14-49ac-9560-afe858a3c756", failure 0
E0427 22:35:48.129032       1 controller.go:966] error syncing claim "f4e0428a-1e14-49ac-9560-afe858a3c756": failed to provision volume with StorageClass "kadalu.external.external-configuration": rpc error: code = InvalidArgument desc = External Storage provided not valid
I0427 22:35:48.129166       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv-ext-kadalu", UID:"f4e0428a-1e14-49ac-9560-afe858a3c756", APIVersion:"v1", ResourceVersion:"3211", FieldPath:""}): type: 'Warning' reason: 'ProvisioningFailed' failed to provision volume with StorageClass "kadalu.external.external-configuration": rpc error: code = InvalidArgument desc = External Storage provided not valid
I0427 22:35:49.129432       1 controller.go:1317] provision "default/pv-ext-kadalu" class "kadalu.external.external-configuration": started
I0427 22:35:49.130063       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pv-ext-kadalu", UID:"f4e0428a-1e14-49ac-9560-afe858a3c756", APIVersion:"v1", ResourceVersion:"3211", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "default/pv-ext-kadalu"
```
- ## csi-attacher
```
I0427 22:38:45.893390       1 reflector.go:515] pkg/mod/k8s.io/client-go@v0.19.0/tools/cache/reflector.go:156: Watch close - *v1.PersistentVolume total 0 items received
I0427 22:39:59.891436       1 reflector.go:515] pkg/mod/k8s.io/client-go@v0.19.0/tools/cache/reflector.go:156: Watch close - *v1.VolumeAttachment total 0 items received
I0427 22:42:06.872172       1 reflector.go:369] pkg/mod/k8s.io/client-go@v0.19.0/tools/cache/reflector.go:156: forcing resync
I0427 22:42:06.872258       1 reflector.go:369] pkg/mod/k8s.io/client-go@v0.19.0/tools/cache/reflector.go:156: forcing resync
```
- ## kaladu-provisioner 
```
[2021-04-27 22:35:48,127] WARNING [volumeutils - 959:check_external_volume] - No host volume found to provide PV
[2021-04-27 22:35:48,127] ERROR [controllerserver - 202:CreateVolume] - External Storage provided not valid
[2021-04-27 22:35:49,133] WARNING [volumeutils - 959:check_external_volume] - No host volume found to provide PV
[2021-04-27 22:35:49,134] ERROR [controllerserver - 202:CreateVolume] - External Storage provided not valid
[2021-04-27 22:35:51,140] WARNING [volumeutils - 959:check_external_volume] - No host volume found to provide PV
[2021-04-27 22:35:51,141] ERROR [controllerserver - 202:CreateVolume] - External Storage provided not valid
[2021-04-27 22:35:55,146] WARNING [volumeutils - 959:check_external_volume] - No host volume found to provide PV
[2021-04-27 22:35:55,147] ERROR [controllerserver - 202:CreateVolume] - External Storage provided not valid
[2021-04-27 22:36:03,154] WARNING [volumeutils - 959:check_external_volume] - No host volume found to provide PV
[2021-04-27 22:36:03,155] ERROR [controllerserver - 202:CreateVolume] - External Storage provided not valid
[2021-04-27 22:36:19,773] ERROR [_server - 445:_call_behavior] - Exception calling application: disk I/O error
Traceback (most recent call last):
  File "/kadalu/lib/python3.8/site-packages/grpc/_server.py", line 435, in _call_behavior
    response_or_iterator = behavior(argument, context)
  File "/kadalu/controllerserver.py", line 148, in CreateVolume
    if not is_hosting_volume_free(ext_volume['name'], pvsize):
  File "/kadalu/volumeutils.py", line 416, in is_hosting_volume_free
    with SizeAccounting(hostvol, mntdir) as acc:
  File "/kadalu/kadalulib.py", line 187, in __enter__
    self.conn = sqlite3.connect(os.path.join(self.mount_path, DB_NAME))
sqlite3.OperationalError: disk I/O error
[2021-04-27 22:36:51,818] INFO [volumeutils - 933:mount_glusterfs_with_host] - mount command failed	 cmd=['/opt/sbin/glusterfs', '--process-name', 'fuse', '-l', '/var/log/gluster/gluster.log', '--volfile-id', 'myvol1', '--volfile-server', '10.13.13.2', '/mnt/external-configuration'] error=[1]  
[2021-04-27 22:36:52,206] ERROR [controllerserver - 202:CreateVolume] - External Storage provided not valid
[2021-04-27 22:37:56,247] INFO [volumeutils - 933:mount_glusterfs_with_host] - mount command failed	 cmd=['/opt/sbin/glusterfs', '--process-name', 'fuse', '-l', '/var/log/gluster/gluster.log', '--volfile-id', 'myvol1', '--volfile-server', '10.13.13.2', '/mnt/external-configuration'] error=[1]  
[2021-04-27 22:37:56,634] ERROR [controllerserver - 202:CreateVolume] - External Storage provided not valid
```
- ## kadalu-logging
```
[2021-04-27 22:36:51.814884 +0000] I [MSGID: 100030] [glusterfsd.c:2697:main] 0-/opt/sbin/glusterfs: Started running version [{arg=/opt/sbin/glusterfs}, {version=2021.03.14}, {cmdlinestr=/opt/sbin/glusterfs --process-name fuse -l /var/log/gluster/gluster.log --volfile-id myvol1 --volfile-server 10.13.13.2 /mnt/external-configuration}] 
[2021-04-27 22:36:51.815699 +0000] E [fuse-bridge.c:6756:init] 0-fuse: Mountpoint /mnt/external-configuration seems to have a stale mount, run 'umount /mnt/external-configuration' and try again.
[2021-04-27 22:36:51.815887 +0000] E [MSGID: 101019] [xlator.c:640:xlator_init] 0-fuse: Initialization of volume failed. review your volfile again. [{name=fuse}] 
[2021-04-27 22:37:56.244654 +0000] I [MSGID: 100030] [glusterfsd.c:2697:main] 0-/opt/sbin/glusterfs: Started running version [{arg=/opt/sbin/glusterfs}, {version=2021.03.14}, {cmdlinestr=/opt/sbin/glusterfs --process-name fuse -l /var/log/gluster/gluster.log --volfile-id myvol1 --volfile-server 10.13.13.2 /mnt/external-configuration}] 
[2021-04-27 22:37:56.245464 +0000] E [fuse-bridge.c:6756:init] 0-fuse: Mountpoint /mnt/external-configuration seems to have a stale mount, run 'umount /mnt/external-configuration' and try again.
[2021-04-27 22:37:56.245523 +0000] E [MSGID: 101019] [xlator.c:640:xlator_init] 0-fuse: Initialization of volume failed. review your volfile again. [{name=fuse}] 
[2021-04-27 22:40:04.680873 +0000] I [MSGID: 100030] [glusterfsd.c:2697:main] 0-/opt/sbin/glusterfs: Started running version [{arg=/opt/sbin/glusterfs}, {version=2021.03.14}, {cmdlinestr=/opt/sbin/glusterfs --process-name fuse -l /var/log/gluster/gluster.log --volfile-id myvol1 --volfile-server 10.13.13.2 /mnt/external-configuration}] 
[2021-04-27 22:40:04.681981 +0000] E [fuse-bridge.c:6756:init] 0-fuse: Mountpoint /mnt/external-configuration seems to have a stale mount, run 'umount /mnt/external-configuration' and try again.
[2021-04-27 22:40:04.682163 +0000] E [MSGID: 101019] [xlator.c:640:xlator_init] 0-fuse: Initialization of volume failed. review your volfile again. [{name=fuse}] 
[2021-04-27 22:44:21.140294 +0000] I [MSGID: 100030] [glusterfsd.c:2697:main] 0-/opt/sbin/glusterfs: Started running version [{arg=/opt/sbin/glusterfs}, {version=2021.03.14}, {cmdlinestr=/opt/sbin/glusterfs --process-name fuse -l /var/log/gluster/gluster.log --volfile-id myvol1 --volfile-server 10.13.13.2 /mnt/external-configuration}] 
[2021-04-27 22:44:21.141847 +0000] E [fuse-bridge.c:6756:init] 0-fuse: Mountpoint /mnt/external-configuration seems to have a stale mount, run 'umount /mnt/external-configuration' and try again.
[2021-04-27 22:44:21.141978 +0000] E [MSGID: 101019] [xlator.c:640:xlator_init] 0-fuse: Initialization of volume failed. review your volfile again. [{name=fuse}] 
[2021-04-27 22:49:21.569186 +0000] I [MSGID: 100030] [glusterfsd.c:2697:main] 0-/opt/sbin/glusterfs: Started running version [{arg=/opt/sbin/glusterfs}, {version=2021.03.14}, {cmdlinestr=/opt/sbin/glusterfs --process-name fuse -l /var/log/gluster/gluster.log --volfile-id myvol1 --volfile-server 10.13.13.2 /mnt/external-configuration}] 
[2021-04-27 22:49:21.570011 +0000] E [fuse-bridge.c:6756:init] 0-fuse: Mountpoint /mnt/external-configuration seems to have a stale mount, run 'umount /mnt/external-configuration' and try again.
[2021-04-27 22:49:21.570089 +0000] E [MSGID: 101019] [xlator.c:640:xlator_init] 0-fuse: Initialization of volume failed. review your volfile again. [{name=fuse}]
```
Anyway i can still mounting the drive from outside of the cluster using `sudo mount -t glusterfs 10.13.13.2:/myvol1 /mnt/myvol1` . So i'm going to try with another network address instead of 10.XX.XX.XX for the gluster_host just to be sure that it is not a network related problem.

**Comment by leelavg** [comment]
@juliosibaja 

Thanks for detailed logs, I'm not comfortable in commenting about failures wrt glusterfs so, will wait for others to pitch in.

> sqlite3.OperationalError: disk I/O error
> [2021-04-27 22:49:21.570011 +0000] E [fuse-bridge.c:6756:init] 0-fuse: Mountpoint /mnt/external-configuration seems to have a stale mount, run 'umount /mnt/external-configuration' and try again.
> [2021-04-27 22:49:21.570089 +0000] E [MSGID: 101019] [xlator.c:640:xlator_init] 0-fuse: Initialization of volume failed. review your volfile again. [{name=fuse}]

IMO above are the major errors that needs to be looked into, I tested on x86_64 but don't have the availability of arm. Will get back at the earliest.

**Comment by amarts** [comment]
All ARM64 issues are now resolved with 0.8.6. Can you upgrade and check if things are working?

**Comment by juliosibaja** [comment]
Hello @amarts!! I can confirm you that the external volume storage and also _internal storage_ are working fine with 0.8.6 ARM64 version in a rpi4 with raspbian and k3s!! Cheers!! 

PD: The overall speed of a replica-2 is relatively slow in my system, but this is a gluster issue anyways.
